### PR TITLE
[hip] Remove deprecated hip APIs, simplify urContext

### PR DESCRIPTION
**Pull request by JackAKirk**
hip amd does not have a native context type other than to support the cuContext type on Nvidia devices. hip plans to completely remove usage of this type anyway, with all corresponding APIs marked deprecated.

This PR future-proofs our hip implementation for when these deprecated APIs become unsupported. It also simplifies the urContext type.

This updates a context test to pass if the interface is unsupported, which cleans up the native_cpu match file that doesn't support certain context interfaces.

### CODE REVIEW

**Comment by JackAKirk** [comment]
intel/llvm tested with these changes here: https://github.com/intel/llvm/pull/14476

File: source/adapters/hip/device.hpp [hunk 1]
```diff
@@ -97,3 +91,20 @@ struct ur_device_handle_t_ {
 };
 
 int getAttribute(ur_device_handle_t Device, hipDeviceAttribute_t Attribute);
+
+namespace {
+/// Scoped Device is used across all UR HIP plugin implementation to activate
+/// the native Device on the current thread. The ScopedDevice does not
+/// reinstate the previous device as all operations in the hip adapter that
```
Comment (frasercrmck) [inline source/adapters/hip/device.hpp:98]: Looks like this was copied from an older comment but we have `HIP` and `hip` in the same comment. I think `HIP` is best?

**Review by EwanC — APPROVED** [review]
command-buffer change LGTM

File: source/adapters/hip/device.hpp (hunk 1 above)
Comment (JackAKirk) [reply source/adapters/hip/device.hpp]: Thanks, I've used HIP consistently now.

**Review by kbenzie — APPROVED** [review]

File: source/adapters/hip/context.hpp [hunk 2]
```diff
@@ -19,35 +19,6 @@
...
-/// <b> HIP context vs UR context </b>
-///
-/// One of the main differences between the UR API and the HIP driver API is
-/// that the second modifies the state of the threads by assigning
-/// \c hipCtx_t objects to threads. \c hipCtx_t objects store data associated
-/// with a given device and control access to said device from the user side.
-/// UR API context are objects that are passed to functions, and not bound
-/// to threads.
-///
-/// Since the \c ur_context_handle_t can contain multiple devices, and a \c
```
Comment (hdelan) [inline source/adapters/hip/context.hpp:36]: Why are you removing this comment? I think it's quite informative

File: source/adapters/hip/context.hpp [hunk 3]
```diff
@@ -76,24 +47,14 @@ struct ur_context_handle_t_ {
...
-  using native_type = hipCtx_t;
-
   std::vector<ur_device_handle_t> Devices;
 
   std::atomic_uint32_t RefCount;
 
   ur_context_handle_t_(const ur_device_handle_t *Devs, uint32_t NumDevices)
-      : Devices{Devs, Devs + NumDevices}, RefCount{1} {
-    for (auto &Dev : Devices) {
-      urDeviceRetain(Dev);
```
Comment (hdelan) [inline source/adapters/hip/context.hpp:88]: Why no longer retaining ur devices?

File: source/adapters/hip/device.hpp [hunk 4]
```diff
@@ -97,3 +91,20 @@ struct ur_device_handle_t_ {
...
+/// reinstate the previous device as all operations in the HIP adapter that
+/// require an active device, set the active device and don't rely on device
+/// reinstation
+class ScopedDevice {
+public:
+  ScopedDevice(ur_device_handle_t hDevice) {
+    if (!hDevice) {
+      throw UR_RESULT_ERROR_INVALID_DEVICE;
+    }
+    UR_CHECK_ERROR(hipSetDevice(hDevice->getIndex()));
```
Comment (hdelan) [inline source/adapters/hip/device.hpp:107]: Nice

**Review by hdelan — APPROVED** [review]
LGTM with a few minor questions.

File: source/adapters/hip/context.hpp (hunk 3 above)
Comment (JackAKirk) [reply source/adapters/hip/context.hpp]: I've added it back.

**Comment by omarahmed1111** [comment]
@JackAKirk Please make sure https://github.com/intel/llvm/pull/14476 is updated and passing CI, so we could merge this. Thanks!

**Comment by kbenzie** [comment]
I've removed *ready to merge* since the intel/llvm PR isn't passing testing and is still awaiting approvals. Once these are in place, add the *ready to merge* label again.

**Comment by JackAKirk** [comment]
This updates a context test to pass if the interface is unsupported, which cleans up the native_cpu match file that doesn't support certain context interfaces.

**Comment by JackAKirk** [comment]
> I've removed _ready to merge_ since the intel/llvm PR isn't passing testing and is still awaiting approvals. Once these are in place, add the _ready to merge_ label again.

Thanks, tests for https://github.com/intel/llvm/pull/14476 pass now and has approval, so I've put back ready to merge. I'll update that PR with the commit hash once this PR is merged.

**Pull request merged**